### PR TITLE
(GH-22) Add tests for Castle Windsor

### LINKED ISSUE #22: Add tests which use Castle Windsor as IoC container
As discussed in #19

### PR DESCRIPTION
**Pull request by marco-bertschi**
Fixes #22

### CODE REVIEW

File: src/BBT.StructureTools.Tests/BBT.StructureTools.Tests.csproj [hunk 1]
```diff
@@ -18,21 +18,23 @@
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
     <DebugType>pdbonly</DebugType>
   </PropertyGroup>
+  <ItemGroup>
+    <None Remove="Compare\ve-B655.tmp" />
+  </ItemGroup>
```
Comment (pascalberger) [inline src/BBT.StructureTools.Tests/BBT.StructureTools.Tests.csproj]: Don't think this should be added here

File: src/BBT.StructureTools.Tests/Compare/ComparerWithObjectAttributeTests.cs [hunk 2]
```diff
@@ -5,48 +5,49 @@
...
     /// <summary>
     /// Test for Comparer infrastructure with object attributes.
     /// </summary>
+    [TestFixture]
+    [TestFixtureSource(typeof(IocTestFixtureSource), "IocContainers")]
     public class ComparerWithObjectAttributeTests
     {
         #region Members, Setup
-        private readonly IComparer<TestClass, ITestCompareIntention> testCandidate;
+        private readonly IComparer<TestClass, ITestCompareIntention> testcandidate;
```
Comment (pascalberger) [inline src/BBT.StructureTools.Tests/Compare/ComparerWithObjectAttributeTests.cs]: Why the change and not using pascal casing (also in other files)?

File: src/BBT.StructureTools.Tests/TestTools/IoC/IIocContainer.cs [hunk 3]
```diff
@@ -5,7 +5,7 @@
     /// <summary>
     /// Abstraction of any IoC container to be used for testing.
     /// </summary>
-    internal interface IIocContainer
+    public interface IIocContainer
```
Comment (pascalberger) [inline src/BBT.StructureTools.Tests/TestTools/IoC/IIocContainer.cs:8]: Why does this need to be public?

File: src/BBT.StructureTools.Tests/TestTools/IoC/IIocContainer.cs [hunk 4]
```diff
@@ -19,6 +19,7 @@ internal interface IIocContainer
         /// <typeparam name="TAbstraction">Abstraction (interface) type.</typeparam>
         /// <typeparam name="TImplementation">Implementation of <typeparamref name="TAbstraction"/>.</typeparam>
         void RegisterSingleton<TAbstraction, TImplementation>()
+            where TAbstraction : class
```
Comment (pascalberger) [inline src/BBT.StructureTools.Tests/TestTools/IoC/IIocContainer.cs:22]: Why this new constraint? And why a class while comment mentions it should be an interface?

File: src/BBT.StructureTools.Tests/TestTools/IoC/IocTestFixtureSource.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    /// needed. See also https://github.com/nunit/docs/wiki/TestFixtureSource-Attribute.
+    /// </summary>
+    internal class IocTestFixtureSource
+    {
+#pragma warning disable SA1401 // Fields should be private
+        /// <summary>
+        /// Returns <see cref="IIocContainer"/> implementations
+        /// for the tested IoC containers.
+        /// </summary>
+        internal static object[] IocContainers =
```
Comment (pascalberger) [inline src/BBT.StructureTools.Tests/TestTools/IoC/IocTestFixtureSource.cs]: Make it a property instead of disabling SA1401

File: src/BBT.StructureTools/Convert/IConvertEngine.cs [hunk 6]
```diff
@@ -5,7 +5,7 @@
     /// </summary>
     /// <typeparam name="TSource">The source to copy from.</typeparam>
     /// <typeparam name="TTarget">The target to copy to.</typeparam>
-    internal interface IConvertEngine<TSource, TTarget>
+    public interface IConvertEngine<TSource, TTarget>
```
Comment (pascalberger) [inline src/BBT.StructureTools/Convert/IConvertEngine.cs]: Why does this need to be public for testing Castle Windsor?

File: src/BBT.StructureTools/Convert/IConvertHelper.cs [hunk 7]
```diff
@@ -5,7 +5,7 @@
     /// <summary>
     /// Helper for the convert implementations.
     /// </summary>
-    internal interface IConvertHelper
+    public interface IConvertHelper
```
Comment (pascalberger) [inline src/BBT.StructureTools/Convert/IConvertHelper.cs]: Why does this need to be public for testing Castle Windsor?

File: src/BBT.StructureTools.Tests/Copy/CopyOperationCreateToManyWithReverseRelationIntTests.cs [hunk 8]
```diff
@@ -1,36 +1,37 @@
-﻿namespace BBT.StructureTools.Tests.Copy
+﻿namespace BBT.Life.LiBase.ITests.General.Services.Tools.Copy
```
Comment (pascalberger) [inline src/BBT.StructureTools.Tests/Copy/CopyOperationCreateToManyWithReverseRelationIntTests.cs]: Wrong namespace (also in other files)

**Review by pascalberger — CHANGES_REQUESTED** [review]

File: src/BBT.StructureTools.Tests/TestTools/IoC/IIocContainer.cs (hunk 4 above)
Comment (marco-bertschi) [reply src/BBT.StructureTools.Tests/TestTools/IoC/IIocContainer.cs:22]: Because Castle Windsor's `Component.For<TAbstraction>` requires a reference type. The `class` constraint in C# doesn't require a class but a reference type.

File: src/BBT.StructureTools.Tests/TestTools/IoC/IIocContainer.cs (hunk 3 above)
Comment (marco-bertschi) [reply src/BBT.StructureTools.Tests/TestTools/IoC/IIocContainer.cs:8]: Because it's used in the constructors of the tests, which themselves are public.

**Comment by marco-bertschi** [comment]
Is preparation work for #45
